open-log/rfcs/0001-storage.md

```diff
@@ -0,0 +1,154 @@
+
+```
+Log Entry:
+  SlateDB Key:   | 0x01 | key (bytes) | sequence (u64) |
+  SlateDB Value: | record value (bytes) |
+```
+
+The `0x01` discriminator is reserved for log entries. Additional record types (e.g., metadata, indexes) may be introduced in future RFCs using different discriminators.
+
+This encoding preserves lexicographic key ordering, enabling key-range scans. Entries for the same key are ordered by sequence number.
+
+#### Variable-Length Key Considerations
+
+With variable-length keys, the boundary between key and sequence number is ambiguous during comparison. Two approaches to handle this:
+
+**Custom comparator** — SlateDB could support a custom comparator that treats the key and sequence number as separate components. This would enable correct ordering without encoding changes.
+
+**Scan filtering** — Without a custom comparator, scans may return entries from adjacent keys when key bytes happen to interleave with sequence number bytes. Results can be filtered by exact key match.
+
+In practice, users are likely to use fixed-length keys, which avoids this issue entirely.
+
+### Sequence Numbers
+
+Sequence numbers are assigned from a global counter that increments on every append. Each key's log is monotonically ordered by sequence number, but the sequence numbers are not contiguous—other keys' appends are interleaved in the global sequence.
+
+This approach simplifies ingestion by avoiding per-key sequence tracking. The trade-off is that sequence numbers do not reflect the count of entries within a key's log.
+
+### Write API
+
+The write API mirrors SlateDB's `write` API. The only supported operation is `append`.
+
+```rust
+struct Record {
+    key: Bytes,
+    value: Bytes,
+}
+
+struct WriteOptions {
+    await_durable: bool,
+}
+
+impl OpenLog {
+    async fn append(&self, record: Record, options: WriteOptions) -> Result<(), Error>;
+    async fn append_batch(&self, records: Vec<Record>, options: WriteOptions) -> Result<(), Error>;
+}
+```
+
+### Scan API
+
+The scan API mirrors SlateDB's scan API. A key is provided along with a sequence number range.
+
+```rust
+struct LogEntry {
+    key: Bytes,
+    sequence: u64,
+    value: Bytes,
+}
+
+// TODO: decide which SlateDB ScanOptions parameters to pass through
+struct ScanOptions {
+}
+
+struct ScanIterator { ... }
+
+impl ScanIterator {
+    async fn next(&mut self) -> Result<Option<LogEntry>, Error>;
+}
+
+impl OpenLog {
+    fn scan(&self, key: Bytes, seq_range: impl RangeBounds<u64>, options: ScanOptions) -> ScanIterator;
+}
+```
+
+### Lag and Count (under consideration)
+
+Without contiguous sequence numbers, computing lag requires additional bookkeeping. The approach under consideration augments SlateDB data structures:
+
+- Each SST tracks the number of records it contains
+- Each SST provides the relative index of a given entry within it
```

Comment:
I'm not certain I understood what you meant by this? do we need to track this for every key+seq?

Reply:
I'd assume so. Kafka by analog stores things like log end offset per partition. This is similar in spirit as I understand it. 

However, we probably need to optimize it if we expect the number of keys to be orders of magnitudes more than kafka partitions.

Reply:
I added some further notes about how this might be implemented. Let me know if it is still not clear.